trace_decoder/src/decoding_mpt.rs

```diff
@@ -0,0 +1,917 @@
+use std::{
+    collections::HashMap,
+    iter::{self, empty, once},
+};
+
+use ethereum_types::{Address, U256, U512};
+use evm_arithmetization_mpt::GenerationInputs;
+use keccak_hash::H256;
+use log::trace;
+use mpt_trie::{
+    nibbles::Nibbles,
+    partial_trie::{HashedPartialTrie, Node, PartialTrie},
+    special_query::path_for_query,
+    trie_ops::TrieOpError,
+    trie_subsets::{create_trie_subset, SubsetTrieError},
+    utils::{IntoTrieKey, TriePath},
+};
+use thiserror::Error;
+
+use crate::{
+    aliased_crate_types::{
+        MptAccountRlp, MptExtraBlockData, MptGenerationInputs, MptTrieInputs, MptTrieRoots,
+    },
+    compact::compact_prestate_processing::MptPartialTriePreImages,
+    decoding::TrieType,
+    processed_block_trace_mpt::{
+        MptProcessedBlockTrace, NodesUsedByTxn, ProcedBlockTraceMptSpec, ProcessedSectionInfo,
+        ProcessedSectionTxnInfo, StateTrieWrites,
+    },
+    trace_protocol::TxnInfo,
+    types::{
+        HashedAccountAddr, HashedNodeAddr, HashedStorageAddr, HashedStorageAddrNibbles,
+        OtherBlockData, TrieRootHash, TxnIdx, EMPTY_ACCOUNT_BYTES_RLPED,
+        ZERO_STORAGE_SLOT_VAL_RLPED,
+    },
+    utils::{hash, optional_field, optional_field_hex, update_val_if_some},
+};
+
+/// Stores the result of parsing tries. Returns a [TraceParsingError] upon
+/// failure.
+pub type MptTraceParsingResult<T> = Result<T, Box<MptTraceParsingError>>;
+
+// TODO: Make this also work with SMT decoding...
+/// Represents errors that can occur during the processing of a block trace.
```

Comment:
I'm not sure why this `TraceParsingError` struct wouldn't be straightforwardly compatible with the SMT design (as all fields are options anyway). Any blockers?

Reply:
Yeah, and actually, in the following PRs, I make this realization and end up merging them. I'll make this change to this PR to reduce noise.